From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Sat, 2 Jul 2022 18:14:53 +0900
Subject: [PATCH] Purpur PR - fix #1025 (#1026)

Original by granny <granny@pl3x.net> (PurpurMC Team)

Copyright (c) 2022 PurpurMC LLC

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 232d26e903b12574121fa4f66e414e72b3d0150a..f1a57f965fc1b8b50437db68faffc9dc29958f3c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -410,31 +410,15 @@ public final class CraftServer implements Server {
         this.pluginManager.registerInterface(JavaPluginLoader.class);
 
         File pluginFolder = this.getPluginsFolder(); // Paper
-        boolean sparkLoaded = false; // Purpur
 
         // Paper start
         if (true || pluginFolder.exists()) {
             if (!pluginFolder.exists()) {
                 pluginFolder.mkdirs();
             }
-            Plugin[] plugins = this.pluginManager.loadPlugins(pluginFolder, this.extraPluginJars());
-            // Paper end
-            for (Plugin plugin : plugins) {
-                try {
-                    String message = String.format("Loading %s", plugin.getDescription().getFullName());
-                    plugin.getLogger().info(message);
-                    plugin.onLoad();
-                    sparkLoaded = sparkLoaded || plugin.getDescription().getName().equalsIgnoreCase("spark"); // Purpur
-                } catch (Throwable ex) {
-                    Logger.getLogger(CraftServer.class.getName()).log(Level.SEVERE, ex.getMessage() + " initializing " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
-                }
-            }
-        } else {
-            pluginFolder.mkdir();
-        }
-
-        // Purpur start
-        if (!sparkLoaded) {
+            // Prismarine start - Purpur PR
+            // Purpur start
+            List<File> extraJars = this.extraPluginJars();
             try {
                 File file = new File("cache", "spark.jar");
                 file.getParentFile().mkdirs();
@@ -456,16 +440,27 @@ public final class CraftServer implements Server {
                     java.nio.file.Files.copy(new java.net.URL("https://sparkapi.lucko.me/download/bukkit").openStream(), file.toPath(), java.nio.file.StandardCopyOption.REPLACE_EXISTING);
                 }
 
-                Plugin spark = this.pluginManager.loadPlugin(file);
-                String message = String.format("Loading %s", spark.getDescription().getFullName());
-                spark.getLogger().info(message);
-                spark.onLoad();
+                extraJars.add(file);
             } catch (Exception e) {
-                getLogger().severe("Failed to download and load spark plugin");
+                getLogger().severe("Purpur: Failed to download and install spark plugin");
                 e.printStackTrace();
             }
+            Plugin[] plugins = this.pluginManager.loadPlugins(pluginFolder, extraJars);
+            // Purpur end
+            // Prismarine end - Purpur PR
+            // Paper end
+            for (Plugin plugin : plugins) {
+                try {
+                    String message = String.format("Loading %s", plugin.getDescription().getFullName());
+                    plugin.getLogger().info(message);
+                    plugin.onLoad();
+                } catch (Throwable ex) {
+                    Logger.getLogger(CraftServer.class.getName()).log(Level.SEVERE, ex.getMessage() + " initializing " + plugin.getDescription().getFullName() + " (Is it up to date?)", ex);
+                }
+            }
+        } else {
+            pluginFolder.mkdir();
         }
-        // Purpur end
     }
 
     // Paper start
